interfaces/builtin/content.go

```diff
@@ -135,15 +135,34 @@ func mountEntry(plug *interfaces.Plug, slot *interfaces.Slot, relSrc string, mnt
 
 func (iface *ContentInterface) ConnectedPlugSnippet(plug *interfaces.Plug, slot *interfaces.Slot, securitySystem interfaces.SecuritySystem) ([]byte, error) {
 	contentSnippet := bytes.NewBuffer(nil)
-	for _, r := range iface.path(slot, "read") {
-		fmt.Fprintln(contentSnippet, mountEntry(plug, slot, r, ",ro"))
-	}
-	for _, w := range iface.path(slot, "write") {
-		fmt.Fprintln(contentSnippet, mountEntry(plug, slot, w, ""))
-	}
-
 	switch securitySystem {
+	case interfaces.SecurityAppArmor:
+		if len(iface.path(slot, "write")) > 0 {
+			fmt.Fprintf(contentSnippet, `
+# In addition to the bind mount, add an AppArmor rule so that
+# snaps may directly access the slot implementation's files. Due
+# to a limitation in the kernel's LSM hooks for AF_UNIX, this rule
+# is needed for using named sockets within the exported
+# directory.
+/var/snap/%s/%s/** mrwklix,
+`, slot.Snap.Name, slot.Snap.Revision)
+		}
+		if len(iface.path(slot, "read")) > 0 {
+			fmt.Fprintf(contentSnippet, `
+# In addition to the bind mount, add an AppArmor rule so that
+# snaps may directly access the slot implementation's files
+# read-only.
+/var/snap/%s/%s/** mrkix,
+`, slot.Snap.Name, slot.Snap.Revision)
```

Comment:
These rules are fine since they only grant the same access to the same files that the bind SecurityMount backend provides, except it allows named sockets to work predictably in the same way.

The 'write' is only needed for named sockets, but thank you for adding the 'read' since it will avoid an inconsistency that may trip people up down the line if people decide to use the slot implementation's path directly.

All that said, the `/var/snap/%s/%s` used in the AppArmor rule is exactly the same as what mountEntry gives for source, with '/** ...' appended. Please update/use resolveSpecialVariable() so that we have AppArmor rules that precisely match the src mount.